### ISSUE TITLE
Comments on "Positioning QR codes in space with HoloLens 2 - building a 'poor man's Vuforia'"

### ISSUE THREAD

**Issue by LocalJoost** [LocalJoost/BlogComments#372]
Original article: https://localjoost.github.io/Positioning-QR-codes-in-space-with-HoloLens-2-building-a-'poor-man's-Vuforia'/

**Comment by Yourismus** [comment]
Hello @LocalJoost :)
Very Awesome project!!

I have tried to get the project onto my HoloLens. Everything worked and the app is able to recognize the qr code.

But somehow the marker and the plane are not displayed.
Have you experienced this issue and could help me with this?

**Comment by LocalJoost** [comment]
Can you please post a little movie of what you are seeing? And what branch are you building, with what scene?

**Comment by Yourismus** [comment]
The scene i am building is Assets/App/Scenes/SampleScnene2.Unity
I have attached a video. The app recognizes the qr code and the link but doesnt display the hologram

https://user-images.githubusercontent.com/75693995/109004218-c06ec400-76a8-11eb-98f3-a6eff6ae1b26.mp4

**Comment by LocalJoost** [comment]
I have no idea. Maybe try holding the QR code a bit higher up to the camera, move your head up and down a little? It works fine here 

https://user-images.githubusercontent.com/4129183/109012812-b651c300-76b2-11eb-94db-f8cf7b557c48.mp4

**Comment by Yourismus** [comment]
Hm okay thank you.
I will test it again.

**Comment by loleklel** [comment]
is there any possibility to get alignment with more than 1 QR?

**Comment by LocalJoost** [comment]
What do you mean exactly? You can of course use two QR codes mark two corners of a larger object. Is that what you mean?

**Comment by loleklel** [comment]
Hi,

I was thinking about an alignment virtual 3D object in real-world objects
using QR codes. Is there maybe another way how to get the perfect
alignment? I just found World Locking Tools, do you have any experience
with this??

thanks in advance,
Robin

út 2. 3. 2021 v 20:14 odesílatel Joost van Schaik <notifications@github.com>
napsal:

**Comment by LocalJoost** [comment]
I am a bit confused. That is exactly what my sample is doing, right? Showing a marker - and an airplane - exactly matched to the position and location of the QR code?

**Comment by loleklel** [comment]
what I need is that the physical object will be alignment with the virtual one but not necessarily with QR code. I want to align the physical object, not the QR code. If I do it with one QR code then it will be not accurate enough...

**Comment by LocalJoost** [comment]
Ah. I see. You want to align _without_ a QR code. I that case, have a look at Azure Spatial Anchors. It's in public preview now if I am not mistaken.

**Comment by AnandMulay** [comment]
Hello Sir, I'm using unity LTS version 2019.4.24, and it's not reading the QR code, other settings are the same as you've mentioned in blog 2, using the latest blog2 git code, which shows updated 14 days back. should I use the older LTS version 2019?

**Comment by LocalJoost** [comment]
Hi @AnandMulay ,

The annoying thing is I get mailed by all kinds of developers claiming they did 'exactly like me' save for that one thing they fail to remember of tell me. You give me _very_ little context to work with. I suppose you are a developer. If this is all the info you get, would _you_ be able to tell something?

I downloaded my code fresh from GitHub and it worked the first time. So what you are going to do is give me more information than that you are only a slightly different version of Unity. 

* First, delete the app from your HoloLens
* Then, please go open the scene2 and turn on the QRDebugDispaler. Build and deploy the app anew
![image](https://user-images.githubusercontent.com/4129183/116082465-b7e03d00-a69b-11eb-939a-a8d5a4d6941c.png)
* Start recording a movie
* Start the app. Give all the consent is asks for.
* Hold the QR code into view. Make sure it shows up to the camera. Move it a little around.

Then send me the movie _and the QR code you are using_.

**Comment by DWishR** [comment]
Just wanted to say thank you. Your finding of `WorldManager.GetNativeISpatialCoordinateSystemPtr()` saved me from having to import any of the XR packages that break my version of MRTK, glad to find it in your code as opposed to the samples. Greatly appreciated!

**Comment by mar-cry** [comment]
Hello sir, what kind of QRcode should be used in your qrcode detection project, meaning what messages should qrcode contain? Thank u very much!

**Comment by LocalJoost** [comment]
For the sample to work, unchanged, it needs to contain "https://localjoost.github.io". You can find a ready-to-use QR contained at the bottom of[ the blog post preceding this one]( https://localjoost.github.io/Reading-QR-codes-with-an-MRTK2-Extension-Service/ ). You can also use any other QR by changing the value here in the project 
![image](https://user-images.githubusercontent.com/4129183/116507697-171e9700-a8c0-11eb-8d73-fa4ac4170383.png)

**Comment by LahlouY** [comment]
Hello @LocalJoost .
We built the project with Master ARM64 and it works perfectly. We wanted to build it with Release mode so we can unplug the hololens from the computer, but in this case the hololens can't scan the QR code (the text block stays empty). any toughts on how to fix this? Thanks

**Comment by LocalJoost** [comment]
Hello @LahlouY ,

I am quite confused. Let me first give you a definitive answer about Release mode - no, it does not work in Release mode, I guess you have to file an issue with Microsoft for that. But I really don't understand why you want to compile in Release mode _so you can unplug the HoloLens from the computer_.  Once you have deployed the app to the HoloLens, you can unplug the HoloLens and simply start the app from All Apps. So I really don't understand what your issue is.

**Comment by LahlouY** [comment]
Indeed, it does work. It just that we had problem to build with master so it didn't work at first . Thanks again

**Comment by LocalJoost** [comment]
> Indeed, it does work. It just that we had problem to build with master so it didn't work at first . Thanks again

Awesome. Great to hear!

**Comment by AMT12345** [comment]
Hello,
Thank you so much for the tutorial. I keep running into the issue "QR tracking not supported". Any suggestions?

**Comment by LocalJoost** [comment]
Do you try to run it from the editor? It only runs on HoloLens2 and only of you follow my build instructions to the letter.

**Comment by AMT12345** [comment]
Yes you are right. I was trying to run it with the Hololens through the Editor. It works now! Thank you so much!

**Comment by xunil82** [comment]
Hi joost,
Maybe you can help me, can I save the real space coordinates of the scanned qr code and just load them at a later time without scanning the qr code again.
Ney

**Comment by LocalJoost** [comment]
Well coordinates are always relative to where HoloLens was when the app started so that would be useless to do. What you might want to do is have a look at spatial Anchors and/or the world locking tools. These allow you to store a spatial location for later reference or even transfer them to other HoloLenses to create a shared frame of reference

**Comment by xunil82** [comment]
Thanks for the tip with the wlt.... sounds promising on the website. spatial anchor I tried, but the data should only be valid as long as the spatial map of the holo are current. Understand it nevertheless not completely and have still too little experience. Somehow I have to create a reference when starting from which I place the data. My approach would be to scan all qr codes if necessary move positions of the targets... save word anchor.... and on restart scan a reference qr and then load world anchor. But this is only theory :) ...must use a local solution, No azur

**Comment by Bastien1612** [comment]
Hello sir,
I have seen that to track and choose your QRCode, you have created the parameter Location Qr value.
![image](https://user-images.githubusercontent.com/33509681/135246011-51cc70ae-80ac-4ceb-9877-363b4bbd4274.png)

And in this picture we can see that the location is set to "https://localjoost.github.io" but there is no QRCode in this link, so how do we settle the QRCode in Unity, do we have to upload it to a link and replace "https://localjoost.github.io" by our new link.

Sorry, if this is redundant with your post on 29 April, but I want to be sure of how it work.

**Comment by LocalJoost** [comment]
I am not quite sure what you mean, but I hope I give you an answer to what I think your question is.

You can basically put ANY text in Location QR Code. Suppose your put "Bastien1612" in "Location QR Code" in Unity. Then you go to basically any OR code generator that can generate a QR code with Bastien1612 in it. For instance [this generator](https://www.generateqrcode.net/create-qr?value=Bastien1612&type=QR.)

Print this code out, or display in on a screen (printing works better) and the jet will appear on that code. 

Does this help?

**Comment by Bastien1612** [comment]
Yes, thanks, that's what I wanted to know.
But can I take again your time with another question which is : Does this work only on Hololens or also on Android and IOS ?

**Comment by LocalJoost** [comment]
Hi,

I have tried it on HoloLens only. It uses a specific Microsoft library (Microsoft.MixedReality.QR) and I don't think that is cross platform. Long story short, I can't answer this question because I don't know.

**Comment by WojciechSwierczykPTC** [comment]
Hi, I spent some time with qr codes tracking library from MS and tested it on hololens 2. At first I was happy that it works perfectly but then I realized that something is wrong. QR codes are automatically cached and after another run of my app, when no qr code is visible, QRCodeWatcher is giving me Added events with outdated location of them, from previous session. I haven't found a way to clear this cache, there is no method anywhere. Have you managed it somehow?

**Comment by LocalJoost** [comment]
Did you try the 'slightly improved version' ? https://github.com/LocalJoost/QRCodeService/tree/slightly_improved_version

**Comment by WojciechSwierczykPTC** [comment]
I have some issues running it but from a look on code I see that you added a check for last time of recognition. I will follow that pattern, thank you for your help :)

**Comment by bwHoloLens2** [comment]
Hello Joost, 

Thank you again for your detailled explanation of the concept, it's actually really helpful and stands out amongst all of Microsoft attempts to convey the idea of QR Tracking.

If i may ask a question, how should one go about changing the model diplayed depending on QR Code being read ?

Thank you in advance.

Using your Unity 2020 project meant for Open XR on Unity 2020.3.24f1

**Comment by LocalJoost** [comment]
@bwHoloLens2 thanks for you kind words. I really appreciate that.
Changing the model is simple. Have a look in scene 2: 
![image](https://user-images.githubusercontent.com/4129183/146823476-57c4f97d-f928-42b2-afd7-9d5919e3774a.png)
there's the model, and a little script 'jetcontroller' that actually listens to the trackercontroller. The whole idea is to act like Vuforia.

**Comment by bwHoloLens2** [comment]
Hi Joost,

Thank you for your quick reply, i was able to make it work for different QR just as you described.

Last question; have you by any chance explored this possibility using your project ?
[Local anchor transfers in Unity](https://docs.microsoft.com/en-us/windows/mixed-reality/out-of-scope/local-anchor-transfers-in-unity)

Thank you in advance

**Comment by LocalJoost** [comment]
Great to hear that, and no I haven't used that. Be aware the QR code only supplies a location and rotation, not an anchor

**Comment by bwHoloLens2** [comment]
Would it be possible to hide the blue figure of the marker in the project without it hiding the airplane model ?
Gave it a shot but my project wasn't displaying the hologram on top of the QR as it was supposed.

To describe my use case;  say i stick the QR code on the surface of some machine, my aim would be to display multiple holograms on variant distances scattered around one QR code. Would this option be possible using your project ? 

Thank you in advance for taking the time to answer me and all of the questions on this comment section.

**Comment by LocalJoost** [comment]
How about just disabling the marker? Or just the sprite render?
![image](https://user-images.githubusercontent.com/4129183/147289004-4b03c6c6-eaa5-41d5-9283-393a8b76a2b1.png)

**Comment by bwHoloLens2** [comment]
Any insight on how i could go on about making this ?

" To describe my use case; say i stick the QR code on the surface of some machine, my aim would be to display multiple holograms on variant distances scattered around one QR code. Would this option be possible using your project ? "

Im thinking about associating multiple trackers with the same QR Code and modifying the function you used to move the hologram to the center of the QR code depending on where i wanna place each hologram around the machine. any advice on how to make it somewhat easier or at least possible ?

**Comment by LocalJoost** [comment]
Uhm, well, how about removing the jet from the JetHolder gameobject and putting your objects in it? 
![image](https://user-images.githubusercontent.com/4129183/147336450-9a2b377b-52f8-441e-b23a-a33408fbc7e9.png)

**Comment by ThibaultPotier** [comment]
Hi
First of, thanks for your work

I'm running into a weird issue. When the app launch, it is asking for some permissions (in my case, microphone use, because of other features) but it is not asking for the photo camera permission. Thus, when trying to use the service, I got a "NotDeclaredByApp" error message.

![image](https://user-images.githubusercontent.com/26235591/157524913-ea85fe70-3cdd-411d-8b91-5ea27e9f476f.png)

I tryed your packaged version (found [here ](https://onedrive.live.com/?authkey=%21AF6qaKGG6JQqgOQ&cid=2406ABD89008AC37&id=2406ABD89008AC37%211461179&parId=2406ABD89008AC37%211461178&o=OneUp))and it asked me for camera picture permission, and worked.

So I tripled check the capabilities under player setting and everything is set as expected in the tutorial (the webcam capabilities is checked). 

The problem clearly seems to be the result of this permission thing, and I Don't know where to signify to the app that it needs this permission. I completely uninstalled the app from hololens and built it again but result is always the same.

Any idea about what is happening here ?

**Comment by LocalJoost** [comment]
Okay, so have checked in "player settings". Does that mean you have checked in Unity? Have you also checked the capabilities of the generated VC++ solution? In the Package.appxmanifest?

**Comment by ThibaultPotier** [comment]
Thanks for your quick response !
I did checked it in Unity. So Now I opened visual studio to look at the manifest and it was unchecked ! 
I checked it, made a new appx and now it works :)

I don't know why I add to change the manifest myself (I did not needed it for other permissions in the past).  

I'm glad it works, thanks again :)

**Comment by LocalJoost** [comment]
You are welcome. I have been in your position more than once. See, in _most_ cases Unity updates the Package.appxmanifest. Unfortunately, it does not _always_ do that. As to why and when, I never been able to find any logic to it. It also happens to other files or assets sometimes. Whenever something like that happens, I just delete the entire generated C++ solution and let Unity generate if from scratch. 

Good luck and happy developing!

**Comment by jeko00** [comment]
Hi Joost
thx for your great work. I was reading your blog about your "poor man’s Vuforia" replacement and in particular the lessons learned section. You say "For continuous tracking of QR codes - like you can do with Vuforia - the internal QR code tracker is simply too slow (still). " what I can confirm. Is there any solution to that except licensing Vuforia e.g. better fine tuning of the tracking (I'm aware of the other points you mentioned in the conclusion section) or using other third party solutions you are aware of? Thx again.

**Comment by LocalJoost** [comment]
No, I am not aware of. I guess there are, but the QR codes fulfill my needs.

**Comment by zhouyuwen123** [comment]
Hello, the work is really good. There are all kinds of problems with the pull down. Can you make a video to pull and pack into Hololens2

**Comment by zhouyuwen123** [comment]
Because I downloaded the QR code scan panel and there is no way to scan the QR code

**Comment by LocalJoost** [comment]
@zhouyuwen123 I have no idea what you are actually asking. I guess English is not your first language (it isn't mine either, btw) but I really don't understand what your problems are. "Pull and pack into HoloLens2" does not ring a bell to me. And no, I am not making videos. That is a lot of work and I am not very fond of making videos. Please read my blogs.  Specifically, read this https://localjoost.github.io/Positioning-QR-codes-in-space-with-HoloLens-2-building-a-'poor-man's-Vuforia'/#conclusion

**Comment by zhouyuwen123** [comment]
I am sorry that I just tried the project. There is no problem with the whole project. I put Hololens2 to scan the QR code and there is no response

**Comment by zhouyuwen123** [comment]
@LocalJoost

**Comment by zhouyuwen123** [comment]
Does it have to use a paper QR code

 @LocalJoost

**Comment by LocalJoost** [comment]
Thank you. That is a question I can answer. No, it doesn't have to be paper but HoloLens has difficulty reading from glossy screens. The reflections from the environment mess up the QR code. So I usually take paper because that is usually matte. If you would take glossy paper (like photo paper) you would have the same problem.

Does this help?

**Comment by zhouyuwen123** [comment]
Ok, thank you. I will try. If there is no accident, there will be no problem

**Comment by zhouyuwen123** [comment]
I am Chinese haha

**Comment by LocalJoost** [comment]
Well nothing wrong with being Chinese, right? I can see lots of blog hits from China, apparently there is a lot happening around HoloLens there 👍

good luck

**Comment by wallacewangchao** [comment]
Hello Mr Joost,

I would like to know, does the size of the QR code matter? Does it influence the positioning of the Marker? If the size does not matter, how the hololens knows the QR code distance? based on 3d scene understanding? Infra light? 

thanks in advance!

**Comment by LocalJoost** [comment]
The size matters, that is the bigger the easier it's recognized. I recommend 10x10 cm. Smaller means you need to get VERY close. How HoloLens exactly pulls it off you will need to ask Microsoft. I know it uses the tracking cameras (unlike Vuforia which uses the Webcam)

**Comment by wallacewangchao** [comment]
> 

Thanks for your info. 
1. yes, the size will influence the recognition rate: https://learn.microsoft.com/en-us/windows/mixed-reality/develop/advanced-concepts/qr-code-tracking-overview#size-of-qr-codes
2. Although not explicitly documented, I seriously doubt that HoloLens infrared camera **CAN** directly detect the size of the the QR code (via image + Time-of-Flight?). I test two markers in different size via [Microsoft's Sample](https://github.com/yl-msft/QRTracking), which can print out the size of the marker (see the picture below). It detects my marker's sizes correctly.

![image](https://user-images.githubusercontent.com/10185985/199112160-a2eaa8b2-c74c-40dd-aba0-c1373398e920.png)

**Comment by lisa2309** [comment]
is it possible to instatiate the marker and the plane instead of setactive?

**Comment by LocalJoost** [comment]
I think so. The prefab should have a SpatialGraphCoordinateSystemSetter, but you can call on that after the instantiation

**Comment by garcalo** [comment]
Hi @LocalJoost! Thanks so much for the amazing work improving microsoft's qr detection. I am interested in qr code tracking in newer versions of unity, particularly unity 2021.3.9; I know you use legacy XR, and understand this is not supported in uniry 2021. I am interested in using openXR implemented by MRTK so that i can run my app through Holographic Remoting. Do you now if you app can be addapted to my needs? If so, do you have any suggestions about the best way to approach this problem? (or anything in particular, any suggestions or remarks are veyr welcomed. Thanks again.

**Comment by LocalJoost** [comment]
@garcalo I am currently in the process of updating a sample to use MRTK3 and newer versions of Unity. Please have a bit of patience. However, I am not sure if it will ever work through Holographic Remoting. I am pretty sure the QR package only runs on HoloLens itself, although someone once wrote me he or she was able to get it working on remoting.

**Comment by garcalo** [comment]
@LocalJoost thanks for the fast answer, I have seen some papers like this one: https://www.mdpi.com/1424-8220/23/4/2168 that claim they use qr tracking library from microsoft through holographic remoting, so i guess it must be possible. Anyhow, is the legacy XR the only issue for the app to work on newer versions of unity or there are more problems like for example version incompatibilities of packages? My plan is to use MRTK 2.8.3, so i thought there would not be any problem coming from there.

**Comment by LocalJoost** [comment]
I just completed updating. You can download the MRTK3 version here https://github.com/LocalJoost/QRCodeService/tree/MRTK3. There is already an OpenXR version based upon MRKT2, see this article https://localjoost.github.io/Upgrading-reading-and-positioning-QR-codes-with-HoloLens-2-to-Unity-2020-+-OpenXR-plugin/

**Comment by garcalo** [comment]
@LocalJoost this is incredibly helpful. Thanks so much for your work. Hope you have a nice day.

**Comment by LocalJoost** [comment]
@garcalo You are welcome. This is what I do. Have a nice day too :)

**Comment by iDavid2619** [comment]
Hi, thanks for the very insightful dive into the QR code implementation on Hololens 2, I'm currently working on a project that aims to connect to a server with information embedded in a QR code (IP and port number)

From the second post, it can be observed that the green surrounding line and the play button to open the link that are default behavors for QR codes are no longer present, is it possible for an explanation on that? I'm looking into validating the QR code and then display a custom "connect" button if valid, but still trying to figure out how, the best approach I've thought of is to replace the "play" button with my custom one and let it trigger a custom script on press

**Comment by LocalJoost** [comment]
@iDavid2619 I am truly sorry - I have no idea what you mean and what your exact question is

**Comment by iDavid2619** [comment]
@LocalJoost Just want to ask about how to change the "play" button to a custom button, and turning off that green border as in the picture in your first post
![shellviewer](https://github.com/LocalJoost/BlogComments/assets/45333770/0dc60785-140e-4993-9873-c6267f1b4bed)

**Comment by LocalJoost** [comment]
You cannot. This is a function of HoloLens itself. For other functionality you will need to write an app

**Comment by iDavid2619** [comment]
But aren't these play buttons gone in the second post where the plane is placed on top of the QR code instead of the green lines and play button

**Comment by LocalJoost** [comment]
Yes. Because in the second post an app is running. Then that play button isn't displayed. It only appears of you run the HoloLens she'll without running an app

**Comment by iDavid2619** [comment]
Thank you so much for clarifying this 😆, not being able to find any documentation is really limiting, is there any official documentation where information such as this can be found? I won't mind if it's filled with APIs, just wish to find a website which clarifies what functions are aviable in the QRCode and when they are available

**Comment by LocalJoost** [comment]
The only other sample I know is the Microsoft one, which I found pretty unclear so I created something myself that I thought to be easier to understand and reuse

**Comment by iDavid2619** [comment]
Thanks for the help, strongly agree about the unclearness, I've read some of your blogs which really helped a whole lot when trying to understand how this stuff works, keep up the good work! 👍

**Comment by LocalJoost** [comment]
Wow. Thanks for the kind words ☺️💪

**Comment by garcalo** [comment]
Hi @LocalJoost! Thanks for your help so far! I am interested in modifying the app so that i can continuously track the QR codes so i can move them around without having to click on a "reset" button every time. Can you help with some indications on how to proceed? 
Thanks in advance!

**Comment by LocalJoost** [comment]
Hi, I am sorry for the late reply, things have been a bit stressful here. You could, of course, always let the reset event go off by itself (after like a second) and then remove the previous indicator only when you found a new one. It takes some fiddling but I think it's doable. However, be aware tracking is not real time, in the sense you can freely move around your QR code and your tracker will follow it - it's not like hand tracking, but more like a stop motion. I don't know how fast it is these days, but I seem to recall it does like 2 positions a second.

**Comment by garcalo** [comment]
Hi @LocalJoost, no worries for the late reply! Thanks for your insight I think it is quite valuable. Again thanks for all the work and for being so helful and always replying to the comments!

**Comment by han1453117821** [comment]
Hello, I do not know why, I use your example project, after installation and running can be recognized normally. When the project was restarted, it was not recognized, no matter how many times the distance Angle was tried. I have opened the project in strict accordance with the release specification on git
